### PR TITLE
Fix GLES3 changing 2D shadow atlas size is broken

### LINKED ISSUE #80148: GLES3 changing 2d shadow atlas size is broken
### Godot version

4.1.1 stable

### System information

Windows 10

### Issue description

this is the default size loaded on startup 1024 which seems fine but it actually have loaded the default [2048] and ignored the size in project settings
image

changing it now to any size will break it 
[2048]
image

[512]
image

### Steps to reproduce

change 2d shadow atlas size in project settings while using compatibility renderer. 

### Minimal reproduction project

_

### LINKED ISSUE #80186: GLES3: Godot game is not loading shadow_atlas_size defined in project settings
### Godot version

4.1.1-stable

### System information

Windows 10

### Issue description

When the project loads in editor, all the settings defined in project settings will be loaded over the default values, but when running the game nearly 95% of them will be ignored resulting in just loading the default settings.

here it's just loading only 4 variables from project setting and i have added the fifth in this pull request https://github.com/godotengine/godot/pull/80151
image

### LINKED ISSUE #80396: GLES3: Color.WHITE or Color(1, 1, 1, 1) is black when used inside _draw() function
### Godot version

4.1.1-stable

### System information

Windows 10

### Issue description

this will draw black rect

```gdscript

func _draw():
	draw_rect(Rect2i(Vector2i.ZERO, img.get_size()), Color(1, 1, 1, 1), false)

func _draw():
	draw_rect(Rect2i(Vector2i.ZERO, img.get_size()), Color.WHITE, false)

```

image

while this will draw white rect

```gdscript

func _draw():
	draw_rect(Rect2i(Vector2i.ZERO, img.get_size()), Color(0.99, 1, 1, 1), false)

```

while this is weird but making circle draws Color(0.99, 1, 1, 1) makes also the _draw_rect to get fixed!

image
 

### Steps to reproduce

use Color.WHITE inside any _draw() function or Color(1, 1, 1, 1) and it will draw black color instead of white color

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by WhalesState**
Fixes https://github.com/godotengine/godot/issues/80148
Fixes https://github.com/godotengine/godot/issues/80186
Fixes https://github.com/godotengine/godot/issues/80396

### CODE REVIEW

File: drivers/gles3/rasterizer_canvas_gles3.cpp [hunk 1]
```diff
@@ -2057,7 +2057,17 @@ void RasterizerCanvasGLES3::set_shadow_texture_size(int p_size) {
 		WARN_PRINT("Attempting to set CanvasItem shadow atlas size to " + itos(p_size) + " which is beyond limit of " + itos(config->max_texture_size) + "supported by hardware.");
```
Comment (lawnjelly) [inline drivers/gles3/rasterizer_canvas_gles3.cpp:2057]: Maybe should be a `return` here if you are going to be changing the buffers below?

**Comment by WhalesState** [comment]
it's already clamping the texture size to the max_texture_size allowed so this is just a friendly warning when you set a larger size

**Comment by WhalesState** [comment]
there are still 2 issues, you are allowed to set a negative size in project settings which will spam error to the compiler, also we need to apply texture filtering if it has smaller size and also the editor is not updating the changes on running the scene and it doesn't respect the size in settings

File: drivers/gles3/rasterizer_canvas_gles3.cpp (hunk 1 above)
Comment (lawnjelly) [reply drivers/gles3/rasterizer_canvas_gles3.cpp:2057]: Ah, yep good point, it should be fine. :+1:

**Comment by lawnjelly** [comment]
> it's already clamping the texture size to the max_texture_size allowed so this is just a friendly warning when you set a larger size

Ah, I missed this I assumed it was doing a reject of the command. Clamping makes sense. However it does look like the check for no change is made above this, so could be moved below the clamping, maybe we could fix this at same time.

**Comment by WhalesState** [comment]
you are right, it's the same in vulkan too, let me edit both

**Comment by WhalesState** [comment]
please don't approve since this is not finished, i will try to track why it doesn't respect the new size on running the scene.

**Comment by lawnjelly** [comment]
> please don't approve since this is not finished, i will try to track why it doesn't respect the new size on running the scene.

If you are still working on it, just convert to draft, there is a (hard to see :grinning: ) button on github on the top right of the page. Then reviewers know to hold off, and it also won't be merged by accident. Then click `ready for review` to move out of draft when finished.

**Comment by WhalesState** [comment]
this was helpful, now we need to put this line somewhere in when the game starts

`RS::get_singleton()->canvas_set_shadow_texture_size(GLOBAL_GET("rendering/2d/shadow_atlas/size"));`

it already exists in the editor _load_from_settings() so where to add it ? in the _init() of the game in `main.cpp`

**Comment by WhalesState** [comment]
why this is closed ?

**Comment by WhalesState** [comment]
please have a look on the `main.cpp` should't we also load all this ? because it will load the defaults from `ResnderingServer` and will ignore the changes in the project settings
image

**Comment by WhalesState** [comment]
@lawnjelly have a look on this, should i open a new issue for adding the other settings in `main.cpp` ?

**Comment by WhalesState** [comment]
while in vulkan it looks like this

[32]
image

[128]
image

[2048]
image

File: drivers/gles3/rasterizer_canvas_gles3.cpp [hunk 2]
```diff
@@ -2048,16 +2048,26 @@ void RasterizerCanvasGLES3::occluder_polygon_set_cull_mode(RID p_occluder, RS::C
...
+	if (p_size == state.shadow_texture_size) {
+		return;
+	}
 	state.shadow_texture_size = p_size;
+
+	if (state.shadow_fb != 0) {
+		glDeleteFramebuffers(1, &state.shadow_fb);
+		glDeleteTextures(1, &state.shadow_texture);
+		glDeleteRenderbuffers(1, &state.shadow_depth_buffer);
+		state.shadow_fb = 0;
```
Comment (clayjohn) [inline drivers/gles3/rasterizer_canvas_gles3.cpp:2066]: The shadow texture needs to be freed using the free_data utility to ensure that we properly track resources
```suggestion
		glDeleteFramebuffers(1, &state.shadow_fb);
		GLES3::Utilities::get_singleton()->texture_free_data(state.shadow_texture);
		glDeleteRenderbuffers(1, &state.shadow_depth_buffer);
		state.shadow_fb = 0;
```

**Review by clayjohn — COMMENTED** [review]
Great work! I just left one suggestion, after fixing that one thing this should be ready to merge

**Comment by WhalesState** [comment]
@clayjohn Thank you, please feel free to review and edit the code since I have no experience with opengl, I also have found another issue on this file.

the default polygon color opengl is black, and here it doesn't update the polygon color if it's white, means drawing with white color will not update the polygon color in opengl but any other color will do, this issue only appears when building without Vulkan and trying to draw any polygon with white color, like a line or a rect in draw function. can i fix this one too ?

image

**Review by clayjohn — APPROVED** [review]
Looks good now! Thanks for working on this. 

As a note for the future, we try really hard to keep separate changes in their own PRs. So, in the future, please open a new PR when you want to fix a separate bug. In this case I won't make you open a new PR because it's literally a two line change. But even for small changes like this it really is better to open a separate PR.

**Comment by WhalesState** [comment]
Got it, Thank you ^^
Should I keep the changes in the header file or remove them after the checks? since they make no difference to those fixes.

**Comment by clayjohn** [comment]
> Should I keep the changes in the header file or remove them after the checks? since they make no difference to those fixes.

You can keep them. I prefer being explicit about initialization anyway.

**Pull request merged**